crates/trigger/src/runtime_config.rs

```diff
@@ -0,0 +1,262 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{Context, Result};
+use serde::Deserialize;
+use spin_config::provider::{env::EnvProvider, vault::VaultProvider};
+use toml;
+
+pub const DEFAULT_STATE_DIR: &str = ".spin";
+const DEFAULT_LOGS_DIR: &str = "logs";
+
+const SPIN_CONFIG_ENV_PREFIX: &str = "SPIN_CONFIG";
+
+const DEFAULT_SQLITE_DB_FILENAME: &str = "sqlite_key_value.db";
+
+/// RuntimeConfig allows multiple sources of runtime configuration to be
+/// queried uniformly.
+#[derive(Debug, Default)]
+pub struct RuntimeConfig {
+    local_app_dir: Option<PathBuf>,
+    files: Vec<RuntimeConfigOpts>,
+    overrides: RuntimeConfigOpts,
+}
+
+impl RuntimeConfig {
+    // Gives more consistent conditional branches
+    #![allow(clippy::manual_map)]
+
+    pub fn new(local_app_dir: Option<PathBuf>) -> Self {
+        Self {
+            local_app_dir: local_app_dir.map(Into::into),
+            ..Default::default()
+        }
+    }
+
+    /// Load a runtime config file from the given path. Options specified in a
+    /// later-loaded file take precedence over any earlier-loaded files.
+    pub fn load_config_file(&mut self, path: impl AsRef<Path>) -> Result<()> {
+        let path = path.as_ref();
+        let bytes = std::fs::read(path)
+            .with_context(|| format!("Failed to load runtime config file {path:?}"))?;
+        let opts = toml::from_slice(&bytes)
+            .with_context(|| format!("Failed to parse runtime config file {path:?}"))?;
+        self.files.push(opts);
+        Ok(())
+    }
+
+    /// Return an iterator of configured spin_config::Providers.
+    pub fn config_providers(&self) -> Vec<BoxedConfigProvider> {
+        // Default EnvProvider
+        let dotenv_path = self.local_app_dir.as_deref().map(|path| path.join(".env"));
+        let env_provider = EnvProvider::new(SPIN_CONFIG_ENV_PREFIX, dotenv_path);
+
+        let mut providers: Vec<BoxedConfigProvider> = vec![Box::new(env_provider)];
+        for provider_config in self
+            .opts_layers()
+            .flat_map(|opts| opts.config_providers.iter())
+        {
+            providers.push(match provider_config {
```

Comment:
It is a nit for now but this is a very beefy match expression to squeeze into a function argument!